Detect challenge binary and libc in `pwn template`

peace-maker opened the pull request:
When the `args.exe` or `args.libc` arguments are missing, look for the files in the current working directory.
This allows to quickly get a template by just running `pwn template` (and maybe `--host`+`--port` for remote instances) and have the arguments filled in automatically. Less typing - more pwning!

This is inspired by https://github.com/io12/pwninit

Related #2276

peace-maker committed:
Detect challenge binary and libc in `pwn template`

When the `args.exe` or `args.libc` arguments are missing, look for the
files in the current working directory.
This allows to quickly get a template by just running `pwn template`
(and maybe --host --port for remote instances).

This is inspired by https://github.com/io12/pwninit

peace-maker committed:
Update CHANGELOG

peace-maker committed:
Add --no-auto argument to disable automatic detection

Allow to keep the old static template behavior.

peace-maker committed:
Add logging about detection outcome

Inform the user that we're looking for the binaries and if we found one.

peace-maker commented:
We could use the `pwnlib.filesystem` abstraction to look for the files on the ssh server too, but I see even more potential in automatically grabbing the used libc from the remote wherever it is instead. That can be done in a separate PR.

Arusekk commented on pwnlib/commandline/template.py:
```suggestion
parser.add_argument('--no-auto', help='Do not automatically detect missing binaries', action='store_false', dest='auto')
```
I suppose we can keep less double negations in code

Arusekk commented on pwnlib/commandline/template.py:
I think we could restrict this a bit

```suggestion
        if not libc and ('libc-' in filename or 'libc.' in filename):
```

Arusekk commented on pwnlib/commandline/template.py:
Idea for the future: more and more challenges ship a `Dockerfile` or even docker-compose. We could use it to detect needed libraries and a binary as well maybe (using API like we do with ADB or shelling out to docker cli).

peace-maker committed:
Be more strict about libc filename detection

Check about common name patterns instead of plainly the presence of "libc" in the name.

Co-authored-by: Arusekk <arek_koz@o2.pl>

peace-maker committed:
Change --no-auto argument to reduce double negation

Co-authored-by: Arusekk <arek_koz@o2.pl>

peace-maker committed:
Merge branch 'dev' into template_detect_files
